Wb data iss1

wbarfuss opened the pull request:
change log:

* pymofa now uses hdf5 databases to store things
* hande call changed
* clean up
* update readme
* update tutorials

To keep in mind:
resaving can be replaced by a second computation of a new handle (see tutorial 1 for example)
==> resaving functions and tutorial 3 can be removed from my side

wbarfuss committed:
pymofa saving h5 DIRTY working

wbarfuss committed:
Added check for computed results

wbarfuss committed:
Add aux function for init to treat raw path

wbarfuss committed:
Added parallel HDF5 storing

wbarfuss committed:
Update README and setup.py

wbarfuss committed:
Update README.rst

wbarfuss committed:
exp handling and tutorial clean up with 'api' changes

wbarfuss committed:
Clean up

wbarfuss committed:
Update README.rst

wbarfuss committed:
Computed tasks tracked in hdf5 db

wbarfuss committed:
Merge branch 'wb_data_iss1' of https://github.com/wbarfuss/modelingframework into wb_data_iss1

wbarfuss commented:
added solution to issue #17

currently running tests on cluster

wbarfuss committed:
Small chance in remaining tasks collection

wbarfuss committed:
Add hdf5 lock clean up func

wbarfuss committed:
Major changes in obtaining computed tasks

wbarfuss committed:
small change in visualiztion

jakobkolb committed:
minor changes

jakobkolb committed:
Moved writing of results to data structure from master to slave. This prevents unpickling errors when too large chunks of data are sent between nodes.

jakobkolb committed:
Moved writing of results to data structure from master to slave. This prevents unpickling errors when too large chunks of data are sent between nodes.

jakobkolb commented on pymofa/experiment_handling.py:
```diff
@@ -135,27 +135,190 @@ def __init__(self, sample_size, parameter_combinations, index, path_raw,
+
+            # METHOD 0) hopefully even faster (incl. restructering of `tasks`)
+            # # # --- obtaining computed tasks (ct)
+            with SafeHDFStore(self.path_raw) as store:
```
This needs more comprehensive inline doc to be comprehensible..

jakobkolb commented on pymofa/experiment_handling.py:
```diff
@@ -135,27 +135,190 @@ def __init__(self, sample_size, parameter_combinations, index, path_raw,
+            rt = rt[rt["__computed"] == False]
+            tasks = rt.drop("__computed", axis=1)
+
+            # method 1: task dataframe
```
This needs a cleanup before merging to master

jakobkolb commented on pymofa/experiment_handling.py:
```diff
@@ -175,63 +338,83 @@ def compute(self, run_func, skipbadruns=False):
+
+                        # store results
+                        # completed runs send thier (task, result) as return
+                        self._obtain_store_function(n_return[0])(n_return[1])
```
I suggest moving this to slave nodes, since sending large packages of data between nodes might lead to unpickling errors (in my experience). Will open issue and commit solution.

jakobkolb commented on pymofa/experiment_handling.py:
```diff
@@ -242,24 +425,85 @@ def compute(self, run_func, skipbadruns=False):
+
+        return mix_names
+
+    def _obtain_store_function(self, task):
```
This function needs more inline documentation to be comprehensive..

jakobkolb reviewed (changes requested):
Nice work!! 
I would like to see minor changes (cleanup and more extensive documentation) before merge to master.

jakobkolb commented:
Also, I approve of the deletion of the resaving function and everything related to it.

jakobkolb committed:
1) Moved the saving of finished tasks from the getter of the save function to the save function itself in _get_store_function() to prevent duplicate entries in computed tasks, that would lead to errors in the _get_computed_tasks() routine.
2) Changed the typecasting in the construction of the dataframe of all tasks in the _get_computed_tasks() routine to correct for weird typecasting in pandas' dataframe construction from dict.

jakobkolb committed:
cosmetic changes for PEP compliance.

jakobkolb committed:
Merge branch 'wb_data_iss1' of https://github.com/wbarfuss/pymofa into wb_data_iss1

# Conflicts:
#	pymofa/experiment_handling.py

jakobkolb committed:
added type hints to public method interfaces of experiment_handling class

jakobkolb committed:
catching SIGTERM issued by slrum 30 seconds prior to termination of run and supressing furter acess to database to make sure, it is closed, when the cluster finally terminates the run

jakobkolb committed:
added some inline comments to explain. This fixes #20

jakobkolb committed:
Merge branch 'wb_data_iss1' of github.com:wbarfuss/pymofa into wb_data_iss1
